iproto: document `box.iproto` module (#3766)
tarantool/doc: doc/dev_guide/reference_capi/box.rst

                                  To override the handler of an unknown request type, use the :ref:`IPROTO_UNKNOWN <internals-iproto-keys-unknown>` type code.

    :param iproto_handler_t handler: IPROTO request handler. To reset the request handler, set the ``handler`` parameter to ``NULL``.
                                     See the full parameter description in the :ref:`Handler function <box_box_iproto_override-handler>` section.

    :param iproto_handler_destroy_t destroy: IPROTO request handler destructor. The destructor is called when the
                                             corresponding handler is removed. See the full parameter description
                                             in the :ref:`Handler destructor function <box_box_iproto_override-destroy>` section.

    :param void* ctx: a context passed to the ``handler`` and ``destroy`` callbacks

    :return: 0 on success
    :return: -1 on error (check :ref:`box_error_last() <c_api-error-box_error_last>`)
    :rtype: number

    See also :ref:`box.iproto.override() <reference_lua-box_iproto_override>`

    **Possible errors:**

>     If the Lua handler throws an exception, the behavior is similar to a remote procedure call.

Reviewer: ```suggestion
    If a Lua handler throws an exception, the behavior is similar to that of a remote procedure call.
```

    The following errors are returned to the client over IPROTO (see `src/lua/utils.h <https://github.com/tarantool/tarantool/blob/dec0e0221e183fa972efa65bb0fb658112f2196f/src/lua/utils.h#L366-L371>`__):

>     *   :errcode:`ER_PROC_LUA` -- the exception is thrown, diagnostic is not set.

Reviewer: ```suggestion
    *   :errcode:`ER_PROC_LUA` -- an exception is thrown from a Lua handler, diagnostic is not set.
```

    *   diagnostics from ``src/box/errcode.h`` -- the exception is thrown, diagnostic is set.

    For details, see `src/box/errcode.h <https://github.com/tarantool/tarantool/blob/master/src/box/errcode.h>`__.

..  _box_box_iproto_override-handler:

    **Handler function**

    The signature of a handler function (the ``handler`` parameter):

    ..  code-block:: c

        enum iproto_handler_status {
                IPROTO_HANDLER_OK,
                IPROTO_HANDLER_ERROR,
                IPROTO_HANDLER_FALLBACK,
        }

        typedef enum iproto_handler_status
        (*iproto_handler_t)(const char *header, const char *header_end,
                            const char *body, const char *body_end, void *ctx);

    where:

    *   ``header``(const char*) -- a MsgPack-encoded header
    *   ``header_end``(const char*) -- end of a header encoded as MsgPack
    *   ``body``(const char*) -- a MsgPack-encoded body
    *   ``header_end``(const char*) -- end of a body encoded as MsgPack

    The handler returns a status code. Possible statuses:

    *  ``IPROTO_REQUEST_HANDLER_OK`` -- success

>     *  ``IPROTO_REQUEST_HANDLER_ERROR`` -- error, diagnostic must be set by handler

Reviewer: I would mention `box_error_set` or `box_error_raise` here.

>     *  ``IPROTO_REQUEST_HANDLER_FALLBACK`` -- fallback to the system handler

Reviewer: ```suggestion
    *  ``IPROTO_REQUEST_HANDLER_FALLBACK`` -- fallback to the system (default) handler
```

..  _box_box_iproto_override-destroy:

    **Handler destructor function**

>     The signature of a destructor function (the ``destroy`` parameter):

Reviewer: Please describe when this destructor is called.

    :return: 0 on success and -1 otherwise. In case of an error user
                could get it via ``box_error_last()``.

..  _box_box_schema_version:

..  c:function:: uint32_t box_schema_version(void)

    Since version :doc:`2.11.0 </release/2.11.0>`.
    Return the database schema version.
    A schema version is a number that indicates whether the :ref:`database schema <index-box-data_schema_description>` is changed.
    For example, the ``schema_version`` value grows if a :ref:`space <index-box_space>` or :ref:`index <index-box_index>`
    is added or deleted, or a space, index, or field name is changed.

    :return: the database schema version
    :rtype: number

    See also :ref:`box.info.schema_version <box_info_schema_version>` and :ref:`IPROTO_SCHEMA_VERSION <internals-iproto-keys-schema_version>`

..  _box_box_session_id:

..  c:function:: uint64_t box_session_id(void)

    Since version :doc:`2.11.0 </release/2.11.0>`.
    Return the unique identifier (ID) for the current session.

>     :return: the unique identifier (ID) for the current session
>     :return: 0 or -1 if there is no session

Reviewer 2: ```suggestion
    :return: the session identifier; 0 or -1 if there is no session
```

    Since version :doc:`2.11.0 </release/2.11.0>`.
    Return the database schema version.
    A schema version is a number that indicates whether the :ref:`database schema <index-box-data_schema_description>` is changed.
    For example, the ``schema_version`` value grows if a :ref:`space <index-box_space>` or :ref:`index <index-box_index>`
    is added or deleted, or a space, index, or field name is changed.

    :return: the database schema version
    :rtype: number

    See also :ref:`box.info.schema_version <box_info_schema_version>` and :ref:`IPROTO_SCHEMA_VERSION <internals-iproto-keys-schema_version>`

..  _box_box_session_id:

..  c:function:: uint64_t box_session_id(void)

    Since version :doc:`2.11.0 </release/2.11.0>`.
    Return the unique identifier (ID) for the current session.

    :return: the unique identifier (ID) for the current session
    :return: 0 or -1 if there is no session
    :rtype: number

    See also :ref:`box.session.id() <box_session-id>`

..  _box_box_iproto_send:

..  c:function:: int box_iproto_send(uint64_t sid, char *header, char *header_end, char *body, char *body_end)

    Since version :doc:`2.11.0 </release/2.11.0>`.

>     Send an :ref:`IPROTO <internals-iproto-format>` packet over the session's socket with the given MsgPack header

Reviewer: I guess you should mention that this feature only works for binary sessions (see [`box.session.type`](https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_session/type/)).

    and body. NB: yields.

    :param uint32_t     sid: IPROTO session identifier (see :ref:`box_session_id() <box_box_session_id>`)
    :param char*     header: a MsgPack-encoded header
    :param char*     header_end: end of a header encoded as MsgPack
    :param char*     body: a MsgPack-encoded body
    :param char*     body_end: end of a body encoded as MsgPack

    :return: 0 on success
    :return: 1 if there is a failure
    :rtype: number

    See also :ref:`box.iproto.send() <reference_lua-box_iproto_send>`

..  _box_box_iproto_override:

..  c:function:: void box_iproto_override(uint32_t request_type, iproto_handler_t handler, iproto_handler_destroy_t destroy, void *ctx)

    Since version :doc:`2.11.0 </release/2.11.0>`.
    Sets an IPROTO request handler with the provided context for the given request type.

>     :param uint32_t     request_type: request type code from the ``iproto_type`` enumeration

Reviewer: I would also describe how `IPROTO_UNKNOWN` works separately. Or I would describe this on the separate page devoted to the Lua API (for `box.iproto.override`) and refer to this page here, i.e., `see details for <box.iproto.override>`.

    :param iproto_handler_t     handler: IPROTO request handler

>     :param iproto_handler_destroy_t destroy: IPROTO request handler destructor

Reviewer: I think you need to describe `iproto_handler_t` and `iproto_handler_destroy_t` separately (i.e., their signature).

>     :param void*    ctx: context passed to handler

Reviewer: This context is actually passed to both the `handler` and `destroy` callbacks (another reason to describe them separately).
